File: erlang/chaos_gm.erl
-module(chaos_gm).
%% Gumowski-Mira chaos

-include_lib("wx/include/wx.hrl").

-export([start/0]).

start()->
	ImgWidth  = 500 ,
	ImgHeight = 500 ,
	
	OvalR     = 1.0 ,
	NumRing   = 15  ,
	
	NumTrajectory         = 100 ,
	NumPointPerTrajectory = 1000 ,
	%%
	ImgType   = "png"   ,
	FileName  = "a.png" ,
	%%
	bitmap:create_picture_file(
		ImgWidth ,
		ImgHeight,
		1        , %% image depth
		ImgType  ,
		FileName ,
		fun(DC)->
			lists:foreach(
			fun(J)->
				IRing = (J rem NumRing) + 1      ,
				R     = (OvalR * IRing)/NumRing  ,
				Theta = random:uniform() * math:pi() * 2.0 , 
				Pt0   = { R*math:cos( Theta ) , R*math:sin( Theta ) } ,
				%%io:format("~w , ~w , ~w~n",[J,Theta,Pt0]) ,
				%%
				loop(NumPointPerTrajectory , Pt0 , DC,{ImgWidth,ImgHeight})
			end ,
			lists:seq(0,NumTrajectory) )
		end
	).


f_0(A,X)-> 
	X_X = X*X , 
	A * X + ( 2*(1-A)*X_X / (1+X_X) ) .

f_1(A,X) ->
	X_X = X*X ,
	A * X + ( 2*(1-A)*X_X / ((1+X_X)*(1+X_X)) ) .

% FROM HTTP://CODEZINE.JP/ARTICLE/DETAIL/327?P=2
f_2(A,X) ->
	X_X = X*X ,
	A * X + ( 2*(1-A)*X / (1+X_X) ) .

% FROM HTTP://CODEZINE.JP/ARTICLE/DETAIL/327?P=2
f_3(A,X) ->
	X_X = X*X ,
	A * X + ( 2*(1-A) / (1+X_X) ) .

g_0({XA,XB},B,Y) ->
	B *Y + XA * (1- XB*Y*Y)*Y.

loop(I,{X,Y},DC,{ImgWidth,ImgHeight})->
	XA    =  0.0005 ,
	XB    =  1.0    ,
	A     = -0.48   ,
	B     =  0.9924 ,
	F     = fun(T)->f_0(A,T) end , 
	G     = fun(T)->g_0({XA,XB},B,T) end ,
	Param = {F,G},
	U     = F(X) ,
	loop(I,{{X,Y},U},Param,DC,{ImgWidth,ImgHeight}).
	
	
loop(I,{{X,Y},U},Param,DC,{ImgWidth,ImgHeight})->
	
	LimitPositive =  1.0e6 ,
	LimitNegative = -1.0e6 ,
	%%
	Ratio     = 20.0       ,
	Offsetx   = ImgWidth /2 ,
	Offsety   = ImgHeight/2 ,
	%%
	%%io:format("(~p,~p)~n",[X,Y]) ,
	if
		I < 0 -> ok;
		X < LimitNegative -> ok ;
		X > LimitPositive -> ok ;
		Y < LimitNegative -> ok ;
		Y > LimitPositive -> ok ;
		true -> 
			IX = erlang:trunc(               Ratio*X + Offsetx   ),
			IY = erlang:trunc( ImgHeight - ( Ratio*Y + Offsety ) ),
			%%
			%%io:format("-->(~p,~p)~n",[IX,IY]) ,
			if
				(IX >= 0) and (IY >=0) and (IX < ImgWidth) and (IY < ImgHeight) ->
					%%io:format("-->drawPoint~n") ,
					wxDC:drawPoint(DC,{IX,IY}) ;
				true -> ok
			end,
			%%
			loop(I-1 , 
				gm_map(Param,{{X,Y},U}) ,
				Param , 
				DC    ,
				{ImgWidth,ImgHeight})
	end.

gm_map({F,G},{{X,Y},U}) ->
	XX = G(Y) + U ,
	UU = F(XX) ,
	YY = UU - X   ,
	{{XX,YY},UU}.

File: ruby/chaos_gm.rb
# Gumowski-Mira attractor
# a.pngに出力

include Java

def createPictureFile(imgWidth , imgHeight , imgType , fileName , &op)
	bi  = java.awt.image.BufferedImage.new( imgWidth , imgHeight , java.awt.image.BufferedImage::TYPE_BYTE_GRAY )
	g   = bi.createGraphics()
	g.setPaint(java.awt.Color::WHITE)
	g.fillRect(0,0,imgWidth,imgHeight)
	g.setColor(java.awt.Color::BLACK)
	
	yield(g)
	
	outFile	= java.io.File.new(fileName)
	javax.imageio.ImageIO.write(bi,imgType,outFile)
end

class Point
	attr_accessor :x , :y
	def initialize(x,y)
		@x = x
		@y = y
	end
end

xa   =  0.0005
xb   =  1.0
a    = -0.48
b    =  0.9924

f_0 = proc{|x|
		x_x = x*x
		a * x + ( 2*(1-a)*x_x / (1+x_x) );
	}

f_1 = proc{|x|
		x_x = x*x
		a * x + ( 2*(1-a)*x_x / ((1+x_x)*(1+x_x)) );
	}
	
# from http://codezine.jp/article/detail/327?p=2
f_2 = proc{|x|
		x_x = x*x
		a * x + ( 2*(1-a)*x / (1+x_x) );
	}
	
# from http://codezine.jp/article/detail/327?p=2
f_3 = proc{|x|
		x_x = x*x
		a * x + ( 2*(1-a) / (1+x_x) );
	}

g_0 = proc{|y|
		b *y + xa * (1- xb*y*y)*y
	}

gm_g = g_0
gm_f = f_0

num_trajectory           = 100
num_point_per_trajectory = 1000

ratio     = 20.0
imgWidth  = 500
imgHeight = 500
offsetx   = imgWidth /2
offsety   = imgHeight/2
imgType   = "png"
fileName  = "a.png"

oval_r    = 0.5
num_ring  = 10

limitPositive = 10.0e6
limitNegative =-limitPositive

createPictureFile(imgWidth , imgHeight , imgType , fileName){ |g|
	rand = java.util.Random.new()
	num_trajectory.times{|j|
		iring = (j % num_ring) + 1 
		r     = (oval_r * iring)/num_ring
		theta = rand.nextDouble * Math::PI * 2
		#
		pt = Point.new( r * Math::cos( theta ) , r * Math::sin( theta ) )
		#
		u = gm_f.call(pt.x)
		num_point_per_trajectory.times{|i|
			x  = pt.x
			y  = pt.y
			break if (limitNegative > x or limitPositive < x  or limitNegative > y or limitPositive < y)
			
			#print pt.x ," , ", pt.y , "\n"
			ix =               ratio*x + offsetx
			iy = imgHeight - ( ratio*y + offsety )
			g.drawLine(ix,iy,ix,iy)
			#
			xx = gm_g.call(y) + u;
			u  = gm_f.call(xx)
			yy = u - x;
			//
			pt.x = xx
			pt.y = yy
		}
	}
}
